src/tests/networking_tests.rs

```diff
@@ -669,6 +670,146 @@ pub mod net_tests {
+        // and also performs clean env setup
+        let _thelock = setup::lock_and_init();
+
+        let cage = interface::cagetable_getref(1);
+
+        let serversockfd = cage.socket_syscall(AF_INET, SOCK_STREAM, 0);
+        let clientsockfd1 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        let clientsockfd2 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        let clientsockfd3 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        let clientsockfd4 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        let clientsockfd5 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        let clientsockfd6 = cage.socket_syscall(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+
+        assert!(serversockfd > 0);
+        assert!(clientsockfd1 > 0);
+        assert!(clientsockfd2 > 0);
+        assert!(clientsockfd3 > 0);
+        assert!(clientsockfd4 > 0);
+        assert!(clientsockfd5 > 0);
+        assert!(clientsockfd6 > 0);
+        let port: u16 = generate_random_port();
+        //binding to a socket
+        let sockaddr = interface::SockaddrV4 {
+            sin_family: AF_INET as u16,
+            sin_port: port.to_be(),
+            sin_addr: interface::V4Addr {
+                s_addr: u32::from_ne_bytes([127, 0, 0, 1]),
+            },
+            padding: 0,
+        };
+        let socket = interface::GenSockaddr::V4(sockaddr); //127.0.0.1
+        assert_eq!(cage.bind_syscall(serversockfd, &socket), 0);
+        assert_eq!(cage.listen_syscall(serversockfd, 5), 0);
+
+        interface::sleep(interface::RustDuration::from_millis(100));
+        assert_eq!(cage.connect_syscall(clientsockfd1, &socket), -115); //EINPROGRESS
+        assert_eq!(cage.connect_syscall(clientsockfd2, &socket), -115);
+        assert_eq!(cage.connect_syscall(clientsockfd3, &socket), -115);
+        assert_eq!(cage.connect_syscall(clientsockfd4, &socket), -115);
+        assert_eq!(cage.connect_syscall(clientsockfd5, &socket), -115);
+        assert_eq!(cage.connect_syscall(clientsockfd6, &socket), -111); 
+
+        assert_eq!(cage.close_syscall(serversockfd), 0);
+        assert_eq!(cage.close_syscall(clientsockfd1), 0);
+        assert_eq!(cage.close_syscall(clientsockfd2), 0);
+        assert_eq!(cage.close_syscall(clientsockfd3), 0);
+        assert_eq!(cage.close_syscall(clientsockfd4), 0);
+        assert_eq!(cage.close_syscall(clientsockfd5), 0);
+        assert_eq!(cage.close_syscall(clientsockfd6), 0);
+
+        assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        lindrustfinalize();
+    }
+
+    //We create a blocking socket to connect to the server and it returns immediately
+    //The blocking socket should not return with success immediately as the server
+    //did not accept the connection
+    #[test]
+    #[ignore]
+    pub fn ut_lind_net_listen_nonblock_then_block() {
```

Comment:
I don't quite understand this test, what's the issue here? Like the connect is supposed to block forever since server is not accepting the socket, but it actually returned immediately?

Reply:
Aren't INET sockets are blocking by default? Hence, as you mentioned, shouldn't the connect sys call block forever since the server is not accepting the socket? However, it returns with success immediately.

Reply:
From what I've read, connect is actually not going to block if server already listened but hasn't accept the connection yet. `connect` can return normally if it goes into the server's listening queue successfully.
btw another suggestion is that it might not be a good idea to add the test that would expect `blocking forever` behavior

Reply:
Yeah thats a good point about adding a test that expects blocking forever behavior. 

Based on the following: https://stackoverflow.com/questions/8258107/does-connect-block-for-tcp-socket , 
so even if the TCP socket is in blocking mode, connect will return with success by being put on the server's listening queue cause the OS takes care of that?

Reply:
nonblocking mode could affect many other functions like accept, send, recv, etc. And in case of connect, there is still some thing get affected. For example, when you try to connect in nonblocking mode, if the server cannot respond immediately (note that this is different from server refuse to respond, in which case the server would explictly send a message back (TCP RST specifically I think)), connect could block until it receives something from server. But in nonblocking mode, it will return after a few milliseconds (must wait some time to determine if server can respond because of network delay) if the server is not responding with INPROGRESS state. Later when the connection is either denied or accept, you can check it using getsockopt syscall. I think the blocking/nonblocking behavior on `connect` syscall, especially on INET is very complex, mostly because you need to balance network delay and nonblocking delay, and it may involve a lot of TCP stuff